### PR TITLE
[libc][math][c23] Add entrypoints and tests for fsqrt{,l,f128}

### PR DESCRIPTION
**Pull request by ghost**
Hello!

My initial pr consists of entry points and the functions `fsqrt`, `fsqrtl`, `fsqrtf128`. These are the instances of the `sqrt` template in `libc/src/__support/FPUtil/generic/`.

Next, I will add the tests.

### CODE REVIEW

File: libc/src/math/fsqrt.h [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_LIBC_SRC_MATH_FSQRT_H
+#define LLVM_LIBC_SRC_MATH_FSQRT_H
+
+#include "src/__support/macros/config.h"
+#include "src/__support/macros/properties/types.h"
```
Comment (lntue) [inline libc/src/math/fsqrt.h:13]: `src/__support/macros/properties/types.h` header is not needed here.

File: libc/src/math/fsqrt.h [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace LIBC_NAMESPACE_DECL {
+
+float fsqrt(double x);
+
+} // namespace LIBC_NAMESPACE_DECL
+
+#endif // LLVM_LIBC_SRC_MATH_F16SQRT_H
```
Comment (lntue) [inline libc/src/math/fsqrt.h:21]: `LLVM_LIBC_SRC_MATH_FSQRT_H`

File: libc/src/math/fsqrtl.h [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_LIBC_SRC_MATH_FSQRTL_H
+#define LLVM_LIBC_SRC_MATH_FSQRTL_H
+
+#include "src/__support/macros/config.h"
+#include "src/__support/macros/properties/types.h"
```
Comment (lntue) [inline libc/src/math/fsqrtl.h:13]: `src/__support/macros/properties/types.h` header is not needed here.

File: libc/src/math/generic/CMakeLists.txt [hunk 4]
```diff
@@ -4189,6 +4189,45 @@ add_entrypoint_object(
...
 )
 
+add_entrypoint_object(
+  fsqrt
+  SRCS
+    fsqrt.cpp
+  HDRS
+    ../fsqrt.h
+  DEPENDS
+    libc.src.__support.macros.properties.types
```
Comment (lntue) [inline libc/src/math/generic/CMakeLists.txt:4199]: `libc.src.__support.macros.properties.types` dependency is not needed

File: libc/src/math/generic/CMakeLists.txt [hunk 5]
```diff
@@ -4189,6 +4189,45 @@ add_entrypoint_object(
...
+)
+
+add_entrypoint_object(
+  fsqrtl
+  SRCS
+    fsqrtl.cpp
+  HDRS
+    ../fsqrtl.h
+  DEPENDS
+    libc.src.__support.macros.properties.types
```
Comment (lntue) [inline libc/src/math/generic/CMakeLists.txt:4212]: `libc.src.__support.macros.properties.types` dependency is not needed

File: libc/test/src/math/smoke/f16sqrtf128_test.cpp [hunk 6]
```diff
@@ -1,4 +1,4 @@
-//===-- Unittests for f16sqrtf128 -----------------------------------------===//
```
Comment (lntue) [inline libc/test/src/math/smoke/f16sqrtf128_test.cpp:1]: this file shouldn't be changed.

File: libc/spec/stdc.td [hunk 7]
```diff
@@ -739,6 +739,10 @@ def StdC : StandardSpec<"stdc"> {
 
           GuardedFunctionSpec<"f16mulf128", RetValSpec<Float16Type>, [ArgSpec<Float128Type>, ArgSpec<Float128Type>], "LIBC_TYPES_HAS_FLOAT16_AND_FLOAT128">,
 
+          FunctionSpec<"fsqrt", RetValSpec<FloatType>, [ArgSpec<DoubleType>]>,
+          FunctionSpec<"fsqrtl", RetValSpec<FloatType>, [ArgSpec<LongDoubleType>]>,
+	  GuardedFunctionSpec<"f16sqrtf128", RetValSpec<FloatType>, [ArgSpec<Float128Type>]>,
```
Comment (lntue) [inline libc/spec/stdc.td:744]: `fsqrtf128`, and the header guard is missing.

File: libc/src/math/fsqrtf128.h [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_LIBC_SRC_MATH_FSQRTF128_H
+#define LLVM_LIBC_SRC_MATH_FSQRTF128_H
+
+#include "src/__support/macros/config.h"
```
Comment (overmighty) [inline libc/src/math/fsqrtf128.h:12]: Should add an `#include` for `float128`.

```suggestion
#include "src/__support/macros/config.h"
#include "src/__support/macros/properties/types.h"
```

File: libc/src/math/generic/CMakeLists.txt [hunk 9]
```diff
@@ -4201,6 +4201,42 @@ add_entrypoint_object(
...
+)
+
+add_entrypoint_object(
+  fsqrtf128
+  SRCS
+    fsqrtf128.cpp
+  HDRS
+    ../fsqrtf128.h
+  DEPENDS
+    libc.src.__support.FPUtil.generic.sqrt
```
Comment (overmighty) [inline libc/src/math/generic/CMakeLists.txt:4235]:
```suggestion
  DEPENDS
    libc.src.__support.macros.properties.types
    libc.src.__support.FPUtil.generic.sqrt
```

File: libc/test/src/math/smoke/fsqrt_test.cpp [hunk 10]
```diff
@@ -0,0 +1,13 @@
+//===-- Unittests for fsqrt -------------------------------------------===//
```
Comment (overmighty) [inline libc/test/src/math/smoke/fsqrt_test.cpp:1]: Nit: formatting.

```suggestion
//===-- Unittests for fsqrt -----------------------------------------------===//
```

File: libc/test/src/math/smoke/fsqrtl_test.cpp [hunk 11]
```diff
@@ -0,0 +1,13 @@
+//===-- Unittests for fsqrtl ------------------------------------------===//
```
Comment (overmighty) [inline libc/test/src/math/smoke/fsqrtl_test.cpp:1]: Nit: formatting.

```suggestion
//===-- Unittests for fsqrtl ----------------------------------------------===//
```

File: libc/spec/stdc.td [hunk 12]
```diff
@@ -739,6 +739,10 @@ def StdC : StandardSpec<"stdc"> {
 
           GuardedFunctionSpec<"f16mulf128", RetValSpec<Float16Type>, [ArgSpec<Float128Type>, ArgSpec<Float128Type>], "LIBC_TYPES_HAS_FLOAT16_AND_FLOAT128">,
 
+          FunctionSpec<"fsqrt", RetValSpec<FloatType>, [ArgSpec<DoubleType>]>,
+          FunctionSpec<"fsqrtl", RetValSpec<FloatType>, [ArgSpec<LongDoubleType>]>,
+	  GuardedFunctionSpec<"fsqrtf128", RetValSpec<FloatType>, [ArgSpec<Float128Type>], "LIBC_TYPES_HAS_FLOAT128">,
```
Comment (overmighty) [inline libc/spec/stdc.td:744]: `fsqrtf128` should be an LLVM libc extension, so it should be moved to libc/spec/llvm_libc_ext.td. The C standard only defines these:

```c
float fsqrt(double x);
float fsqrtl(long double x);
_FloatM fMsqrtfN(_FloatN x); // M < N
```

**Review by overmighty — COMMENTED** [review]
Could you please update the implementation status table in the docs? Since `fsqrtf128` is an LLVM libc extension, it should have a `|check|\*` and not just a `|check|` like the others.

**Comment by overmighty** [comment]
I've changed the PR title to better reflect the fact that it adds multiple functions/entrypoints, if you don't mind.

File: libc/src/math/fsqrtf128.h (hunk 8 above)
Comment (Jobhdez) [reply libc/src/math/fsqrtf128.h:12]: tue told me to remove `#include "src/__support/macros/properties/types.h"`

File: libc/src/math/generic/CMakeLists.txt (hunk 9 above)
Comment (Jobhdez) [reply libc/src/math/generic/CMakeLists.txt:4235]: tue told me remove `libc.src.__support.macros.properties.types`

**Comment by Jobhdez** [comment]
> I've changed the PR title to better reflect the fact that it adds multiple functions/entrypoints, if you don't mind.

its ok

File: libc/src/math/fsqrtf128.h (hunk 8 above)
Comment (Jobhdez) [reply libc/src/math/fsqrtf128.h:12]: @overmighty  let me know whether to add this again please. The tests ran and passed without it.

Comment (Jobhdez) [reply libc/src/math/fsqrtf128.h:12]: nevermind. i think i do need it

File: libc/src/math/generic/CMakeLists.txt (hunk 9 above)
Comment (Jobhdez) [reply libc/src/math/generic/CMakeLists.txt:4235]: never mind

**Review by lntue — APPROVED** [review]

**Pull request merged**